provider/ns1.go

```diff
@@ -0,0 +1,295 @@
+	token, ok := os.LookupEnv("NS1_APIKEY")
+	if !ok {
+		return nil, fmt.Errorf("NS1_APIKEY environment variable is not set")
+	}
+
+	apiClient := api.NewClient(client, api.SetAPIKey(token))
+
+	provider := &NS1Provider{
+		client:       NS1DomainService{apiClient},
+		domainFilter: config.DomainFilter,
+		zoneIDFilter: config.ZoneIDFilter,
+	}
+	return provider, nil
+}
+
+// Records returns the endpoints this provider knows about
+func (p *NS1Provider) Records() ([]*endpoint.Endpoint, error) {
+	zones, err := p.zonesFiltered()
+	if err != nil {
+		return nil, err
+	}
+
+	var endpoints []*endpoint.Endpoint
+
+	for _, zone := range zones {
+
+		// TODO handle Header Codes
+		zoneData, _, err := p.client.GetZone(zone.String())
+		if err != nil {
+			return nil, err
+		}
+
+		for _, record := range zoneData.Records {
+			if supportedRecordType(record.Type) {
+				endpoints = append(endpoints, endpoint.NewEndpointWithTTL(
+					record.Domain,
+					record.Type,
+					endpoint.TTL(record.TTL),
+					record.ShortAns...,
+				),
+				)
+			}
+		}
+	}
+
+	return endpoints, nil
+}
+
+// ns1BuildRecord returns a dns.Record for a change set
+func ns1BuildRecord(zoneName string, change *ns1Change) *dns.Record {
+	record := dns.NewRecord(zoneName, change.Endpoint.DNSName, change.Endpoint.RecordType)
+	for _, v := range change.Endpoint.Targets {
+		record.AddAnswer(dns.NewAnswer(strings.Split(v, " ")))
+	}
+	// set detault ttl
+	var ttl = ns1DefaultTTL
+	if change.Endpoint.RecordTTL.IsConfigured() {
+		ttl = int(change.Endpoint.RecordTTL)
+	}
+	record.TTL = ttl
+
+	return record
+}
+
+// ns1SubmitChanges takes an array of changes and sends them to NS1
+func (p *NS1Provider) ns1SubmitChanges(changes []*ns1Change) error {
+	// return early if there is nothing to change
+	if len(changes) == 0 {
+		return nil
+	}
+
+	zones, err := p.zonesFiltered()
+	if err != nil {
+		return err
+	}
+
+	// separate into per-zone change sets to be passed to the API.
+	changesByZone := ns1ChangesByZone(zones, changes)
+	for zoneName, changes := range changesByZone {
+		for _, change := range changes {
+			record := ns1BuildRecord(zoneName, change)
+			logFields := log.Fields{
+				"record": record.Domain,
+				"type":   record.Type,
+				"ttl":    record.TTL,
+				"action": change.Action,
+				"zone":   zoneName,
+			}
+
+			log.WithFields(logFields).Info("Changing record.")
+
+			if p.dryRun {
+				continue
+			}
+
+			switch change.Action {
+			case ns1Create:
+				_, err := p.client.CreateRecord(record)
+				if err != nil {
+					return err
+				}
+			case ns1Delete:
+				_, err := p.client.DeleteRecord(zoneName, record.Domain, record.Type)
+				if err != nil {
+					return err
+				}
+			case ns1Update:
+				_, err := p.client.UpdateRecord(record)
+				if err != nil {
+					return err
+				}
+			}
+		}
+	}
+	return nil
+}
+
+// Zones returns the list of hosted zones.
+func (p *NS1Provider) zonesFiltered() ([]*dns.Zone, error) {
+	// TODO handle Header Codes
```

Comment:
Same comment as before: can we resolve this TODO instead of leaving it here?

Reply:
As before, this TODO is to handle the edge case where NS1 paginates the response.  In this case, this will only occur with 2500+ zones in a single account and can be mitigated by using an API key that only has access to a subset of zones.

Again, this would be cleaner to resolve after we get pagination supported via the ns1-go sdk